compiler/rustc_mir/src/transform/function_references.rs

```diff
@@ -0,0 +1,158 @@
+use rustc_hir::def_id::DefId;
+use rustc_middle::mir::visit::Visitor;
+use rustc_middle::mir::*;
+use rustc_middle::ty::{self, Ty, TyCtxt};
+use rustc_session::lint::builtin::FUNCTION_REFERENCES;
+use rustc_span::Span;
+use rustc_target::spec::abi::Abi;
+
+use crate::transform::{MirPass, MirSource};
+
+pub struct FunctionReferences;
+
+impl<'tcx> MirPass<'tcx> for FunctionReferences {
+    fn run_pass(&self, tcx: TyCtxt<'tcx>, _src: MirSource<'tcx>, body: &mut Body<'tcx>) {
+        let source_info = SourceInfo::outermost(body.span);
+        let mut checker = FunctionRefChecker {
+            tcx,
+            body,
+            potential_lints: Vec::new(),
+            casts: Vec::new(),
+            calls: Vec::new(),
+            source_info,
+        };
+        checker.visit_body(&body);
+    }
+}
+
+struct FunctionRefChecker<'a, 'tcx> {
+    tcx: TyCtxt<'tcx>,
+    body: &'a Body<'tcx>,
+    potential_lints: Vec<FunctionRefLint>,
+    casts: Vec<Span>,
+    calls: Vec<Span>,
+    source_info: SourceInfo,
+}
+
+impl<'a, 'tcx> Visitor<'tcx> for FunctionRefChecker<'a, 'tcx> {
+    fn visit_basic_block_data(&mut self, block: BasicBlock, data: &BasicBlockData<'tcx>) {
+        self.super_basic_block_data(block, data);
+        for cast_span in self.casts.drain(..) {
```

Comment:
I don't think I have seen a lint before that would do anything at the basic block level... why is this needed? There should be comments in the code explaining anything that is unusual, and this definitely is. :)